### PR TITLE
[image_picker] iOS: save image to the correct type based on its original type and copy over exif data from the original image

### PR DESCRIPTION
**Pull request by cyanglaz**
## Description

Previously, our implementation checks the alpha of the UIImage object to determine the type (png or jpeg). This PR refactored the code to actually check the original image file's type to determine the type it should be saved. Which makes it safer.

While doing so, we also copied the exif data from the original image, now the picked image has the correct exif data.

Since iOS doesn't support converting UIImage to NSData for types other than PNG or JPEG, we are currently only supporting these 2 types. (If anyone knows how to convert UIImage to other types:gif or tiff, we can add them as well)

XCTests target is also added in this PR. So we can start adding iOS unit tests in image picker.

This is a follow up of https://github.com/flutter/plugins/pull/742
And this PR is the first step to achieve what we were trying to do in https://github.com/flutter/plugins/pull/866.

Thanks to @ko2ic for inspiration of some of the works in this PR.

We will need follow up PR to support GIF if possible, as well as copying GPS data over.

## Related Issues

https://github.com/flutter/flutter/issues/31751
https://github.com/flutter/flutter/issues/17947
https://github.com/flutter/flutter/issues/12921

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by cyanglaz** [comment]
@ko2ic This PR is inspired by your PR in https://github.com/flutter/plugins/pull/866
Do you mind review this? :)

File: packages/image_picker/ios/Classes/ImagePickerMetaDataUtil.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
++ (NSData *)updateEXIFData:(NSDictionary *)exifData toImage:(NSData *)imageData;
+
+// Converting UIImage to a NSData with the type proveide.
+//
+// The quality is for JPEG type only, it defaults to 1. It throws exception if setting a non-nil
+// quality with type other than FlutterImagePickerMIMETypeJPEG. Converting UIImage to
+// FlutterImagePickerMIMETypeGIF or FlutterImagePickerMIMETypeTIFF is not supported in iOS. This
+// method throws exception if trying to do so.
++ (nonnull NSData *)convertImage:(nonnull UIImage *)image
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/image_picker/ios/Classes/ImagePickerMetaDataUtil.h]: I think we can add the gif handling in a separate PR. I also want to take a deeper dive into the implementation to see if there is a way to support scaling gif. If not, then we need to decide how to handle when people passed the scaling factors into the API.

**Comment by cyanglaz** [comment]
@ko2ic Thanks for doing this. I updated the PR with you comments! I also added the exif in this PR.

**Comment by ko2ic** [comment]
@cyanglaz I tried it, but the time of Exif is different.
Timezone seems to be related, but I did not understand the regularity.

**Comment by cyanglaz** [comment]
@ko2ic That's strange, all I did was just copying everything over, I will take a look again.

**Comment by cyanglaz** [comment]
@ko2ic I just checked and it seems working for me? Could you post the steps of your testing here?

**Comment by ko2ic** [comment]
@cyanglaz It works without problem in unit test but I get problems when I run it on the simulator.

I put a breakpoint on line 31 in ImagePickerPhotoAssetUtil.m and I confirmed createDate.
The selected image is an image in the simulator.

Please check the image below.

create_date

**Comment by cyanglaz** [comment]
@ko2ic Thank you. I tested with the steps you provided and I found out that the time that shows up in photo app in the simulator does depend on the time zone of your computer. Somehow the time in the metadata is based on Pacific time and it just matched my local time. If you set your mac to pacific timezone, and reopen the photo app in the simulator, you will see the time match the time in the exif.

**Comment by ko2ic** [comment]
@cyanglaz I do not mind if this is a problem with mac and simulator.

**Comment by cyanglaz** [comment]
@ko2ic Thanks! Other than that, do you see any issues or improvements that can be done in the code?

**Comment by ko2ic** [comment]
@cyanglaz No problem. Please go ahead and proceed with it.

**Pull request merged**